core/src/history/events.rs

```diff
@@ -118,6 +115,40 @@ impl EventRegistry {
             .into_iter()
             .map(|(_, Value { event, batch_id })| (event, batch_id))
     }
+
+    /// Returns an iterator containing all events up to and including the
+    /// specified batch ID.
+    pub fn events_until_batch(
+        &self,
+        batch_id: impl Into<BatchId>,
+    ) -> impl Iterator<Item = (&'_ batch_exchange::Event, BatchId)> + '_ {
+        self.events_for_batch_range(BatchId(0), batch_id.into())
+    }
+
+    /// Returns an iterator containing all events that occured in the specified
+    /// batch ID.
+    pub fn events_for_batch(
+        &self,
+        batch_id: impl Into<BatchId>,
+    ) -> impl Iterator<Item = (&'_ batch_exchange::Event, BatchId)> + '_ {
+        let batch_id = batch_id.into();
+        self.events_for_batch_range(batch_id, batch_id)
+    }
+
+    /// Returns an iterator containing all events in the specified batch range.
+    /// Note that the range is inclusive, which means events that were emitted
+    /// on the starting and ending batches are contained in the iterator.
+    fn events_for_batch_range(
+        &self,
+        start_batch_id: BatchId,
+        end_batch_id: BatchId,
+    ) -> impl Iterator<Item = (&'_ batch_exchange::Event, BatchId)> + '_ {
+        self.events
+            .values()
+            .skip_while(move |entry| entry.batch_id < start_batch_id)
+            .take_while(move |entry| entry.batch_id <= end_batch_id)
+            .map(|entry| (&entry.event, entry.batch_id))
```

Comment:
This is possible in a more efficient way with `BTreeMap::range`. Then we might as well accept a range as the parameter instead of start and end batch ids.
I'm also wondering if it wouldn't be better to give direct access to `self.events` instead of going through a layer between. Could even provide the `map(|entry| (&entry.event, entry.batch_id))` as a free function if this is what most consumers want. This approach has less encapsulation which is a negative if we want to refactor the internals in the future but a positive in terms of amount of code needed just for the encapsulation right now. I'm fine with either approach.

Reply:
>  is possible in a more efficient way with `BTreeMap::range`.

Unfortunately, this does not work for values, only for keys :disappointed:. I also couldn't find any way to do a binary search on values (despite them being ordered in our binary tree).

 > I'm also wondering if it wouldn't be better to give direct access to self.events instead [...] This approach has less encapsulation

Personally, I prefer more encapsulation. Specifically, I imagine their may be some invariants that are upheld by `handle_event_data` that would be possible to break otherwise. Additionally, it would require making the key and value structs public as well.

> Could even provide the map(|entry| (&entry.event, entry.batch_id)) as a free function

I think having an `events` iterator like this makes a lot of sense... I will update.